apps/meteor/ee/server/local-services/reads/service.ts

```diff
@@ -0,0 +1,48 @@
+import { Reads } from '@rocket.chat/models';
+import type { ISubscription } from '@rocket.chat/core-typings';
+import { ServiceClassInternal } from '@rocket.chat/core-services';
+
+import type { IReadsService } from '../../sdk/types/IReadsService';
+import { Messages, Subscriptions } from '../../../../app/models/server';
+import { ReadReceipt } from '../../lib/message-read-receipt/ReadReceipt';
+import { MAX_ROOM_SIZE_CHECK_INDIVIDUAL_READ_RECEIPTS } from '../../lib/constants';
+
+export class ReadsService extends ServiceClassInternal implements IReadsService {
+	protected name = 'reads';
+
+	async readThread(userId: string, tmid: string): Promise<void> {
+		const read = await Reads.findOneByUserIdAndThreadId(userId, tmid);
+
+		const threadMessage = Messages.findOneById(tmid, { projection: { ts: 1, tlm: 1, rid: 1 } });
+		if (!threadMessage || !threadMessage.tlm) {
+			return;
+		}
+
+		await Reads.updateReadTimestampByUserIdAndThreadId(userId, tmid);
+		ReadReceipt.storeThreadMessagesReadReceipts(tmid, userId, read?.ls || threadMessage.ts);
+
+		// doesn't mark as read if not all room members have read the thread
+		const subscriptions = await Subscriptions.findByRoomId(threadMessage.rid, {
```

Comment:
You should retrieve the subscriptions only if they are less then `MAX_ROOM_SIZE_CHECK_INDIVIDUAL_READ_RECEIPTS`. So you can do a count first and retrieve them later, you can exclude the archived ones in the query as well